update to new linter API

dirk-thomas opened the pull request:
The patch makes the package use the new linter API and addresses #17, #23, #24, #26.

Besides making it work again the regular expression has been updated to match the three-lines-per-error output of `xmllint`. It uses the indentation of the `^` marker on the third line to determine the column index of the error.

I have only used this patch on Linux. It would great if someone can confirm that it works on Windows too.

Arcanemagus commented on lib/init.coffee:
```diff
@@ -1,8 +1,56 @@
-      default: ''
-
+      title: 'xmllint Executable Path'
+      default: '/usr/bin/xmllint'
```
Just go with `xmllint`, the user's path should figure out the rest.

dirk-thomas replied:
Ok, changed.

Arcanemagus commented on lib/init.coffee:
```diff
@@ -1,8 +1,56 @@
+    @subscriptions.add atom.config.observe 'linter-xmllint.xmllintExecutablePath',
+      (executablePath) =>
+        @executablePath = executablePath
+    console.log 'activate: ' + @executablePath
```
Although I'm not strictly against this, most providers have moved away from sending a message to the console on activate.

If you keep this use the linter name, not the executable path.

dirk-thomas replied:
Makes sense. I just replicated what I found in other packages. I removed both `console.log` calls.

Arcanemagus commented on lib/init.coffee:
```diff
@@ -1,8 +1,56 @@
+    @subscriptions.dispose()
+
+  lintFile: (filePath)->
+    helpers = require 'atom-linter'
```
I would require this at the start of the file, it may be cached but this is a blocking operation.

dirk-thomas replied:
Moved.

Arcanemagus commented on lib/init.coffee:
```diff
@@ -1,8 +1,56 @@
+    return helpers.exec(@executablePath, params, {stream: 'stderr'})
+
+  parseMessages: (output) ->
+    XRegExp = require('xregexp').XRegExp
```
Same for this, no reason not to move it to the start of the file.

dirk-thomas replied:
Moved.

Arcanemagus commented on lib/init.coffee:
```diff
@@ -1,8 +1,54 @@
+  lintFile: (filePath)->
+    helpers = require 'atom-linter'
+    params = ['--noout', filePath]
+    return helpers.exec(@executablePath, params, {stream: 'stderr'})
```
This may or may not make a difference with this linter, but in many others specifying the `cwd` here solves a lot of issues. I'm not familiar with this linter so this may be a non-issue here.

(Searching for dependent files? Config files?)

dirk-thomas replied:
Afaik `xmllint` has no config file nor does it use the cwd anywhere. Also this package didn't set a cwd before so I would rather not set one if there is no reason to do so.

Arcanemagus commented:
While you are at it, you might want to:
* Add a `coffeelint.rc`
* Automatically install the `linter` package using [`atom-package-deps`](https://github.com/steelbrain/package-deps)

Arcanemagus commented on lib/init.coffee:
```diff
@@ -1,8 +1,54 @@
+    @subscriptions.add atom.config.observe 'linter-xmllint.xmllintExecutablePath',
+      (executablePath) =>
+        @executablePath = executablePath
+    @executablePath = '/usr/bin/xmllint'
```
Remove this line, any point it would have is handled by the default value.

dirk-thomas replied:
Sorry, that was left over from debugging. I should have checked the diff more thorough before creating the PR. Removed.

Arcanemagus commented:
Also the README.md needs to be updated.

Arcanemagus commented on lib/init.coffee:
```diff
@@ -1,8 +1,56 @@
+      default: 'xmllint'
   activate: ->
-    console.log 'activate linter-xmllint'
+    require('atom-package-deps').install('linter-xmllint')
```
Since this is a new package you should depend on version 3 and just leave the parameter for `install()` empty.

dirk-thomas replied:
Done in b9435222dfb14fa036aed63264ca195ae2168162

dirk-thomas commented:
Can you give me a link to more information about `coffeelint.rc`. This is my first contribution to an Atom package and I have no idea what it is.

I used `atompackage-deps` as I have seen it in other packages: fcf9784918cb2ec4d38e1ed5312d77cef87cfd59 Is that the correct way?

Can you please elaborate which part of the README are you referring to? The only thing which I see is `It will lint on edit and/or on save` which should be changed to `only on save`.

Arcanemagus commented:
Sorry, `coffeelint.json`, I would just copy the one from `linter`, [here](https://github.com/Arcanemagus/linter/blob/master/coffeelint.json) (And make sure you install `linter-coffeelint` so you can see if your code has any issues that it would expose :wink:).

Looks like you already found the new way of using `atom-package-deps` :+1:.

The readme currently mentions needing to install `linter`, which should be handled automatically for the user now. It also claims the linter will work on "edit and/or on save", when it really only will work on save.

Arcanemagus commented:
Also you should probably just remove the donation section of the README, and cut out the list of other linters (If you want to keep that in there, just change it to a link to http://atomlinter.github.io/.

Arcanemagus commented:
Thanks for taking this on btw, would you be interested in becoming this package's maintainer?

Arcanemagus commented:
Probably should make Windows not look like a sub-section of Linux in the README.

dirk-thomas commented:
Based on the coffee linter I fixed on style issue. I also updated the README.

Thanks, but for now I don't want to jump in and become a maintainer immediately. But I am planning to also look into DTD / XSD validation. May be when I work on the next change (and have worked a bit more with coffeescript) I may become the maintainer?

dirk-thomas commented:
But there is one problem now. (I think that was also the reason why I still had the debug line setting the `@executablePath` to a fixed value.) When I use the package it can;t invoke the command line since `@executablePath` is `undefined`. Do I have to explicitly *get* the default value somehow?

Arcanemagus commented on lib/init.coffee:
```diff
@@ -1,8 +1,55 @@
+    require('atom-package-deps').install()
+
+    @subscriptions = new CompositeDisposable
+    @subscriptions.add atom.config.observe 'linter-xmllint.xmllintExecutablePath',
```
This should be observing `'linter-xmllint.executablePath'`.

dirk-thomas replied:
Thanks. That was obvious (after you pointed it out :wink:).

dirk-thomas commented:
I think I addressed all the comments?

One further question: what can I do in order for this linter to be listed on http://atomlinter.github.io/ (once this patch has been released and the package actually works again)?

Arcanemagus commented:
To get it listed under there you should file a PR over on [the repo](https://github.com/AtomLinter/atomlinter.github.io/pulls) that that is maintained under.

I'm firing up a linux VM to test this on as I don't particularly feel like going through the steps to install this on Windows right now.

Arcanemagus commented:
Can you specify a linter name? (provideLinter -> name attribute)

As soon as that's in there I'll merge this as it's working fine for me. :smile:

dirk-thomas commented:
I added the name: https://github.com/AtomLinter/linter-xmllint/pull/27/files#diff-801ad5e30099db03216b57217be02297R50

Thank you for iterating on this ticket that fast.

Arcanemagus commented:
Thanks for updating this!

Arcanemagus commented:
Released as v1.0.0 :fireworks:.

dirk-thomas commented:
@Arcanemagus While only working with Atom for a short time I am already seeing the opportunity to contribute to every bit in an easy way. I started writing my first own package extending [build](https://atom.io/packages/build) and while doing that created a PR for the [settings-view](https://atom.io/packages/settings-view). If you would still offer me maintainer status of this package I would be happy to take over the role for this package. I think it is a good way of getting starting in the Atom community. I guess I would have full access to the repo then and have to (learn to) do releases etc then?

I already spent some time writing specs for the existing functionality of `linter-xmllint` and would like to enable Travis integration for the repo to reduce the chances of regressions before adding more features.

Arcanemagus commented:
@steelbrain mind creating a team and adding him to it?

steelbrain commented:
@dirk-thomas You're in! Check your emails!

dirk-thomas commented:
Thanks, I got the invitation. But it seems like I still can neither manage any of the tickets nor register the Travis CI for the repo.

Arcanemagus commented:
Looks like you haven't accepted it? You still don't have the badge showing you have access.

dirk-thomas commented:
I think I did accept the invitation. I see myself in the team: https://github.com/orgs/AtomLinter/teams/linter-xmllint

But the team itself seems to have no repo associated to it?

Arcanemagus commented:
@steelbrain did the team get associated with this repo?

steelbrain commented:
@Arcanemagus Sorry man! Missed it

Arcanemagus commented:
@dirk-thomas Since it sounds like you are new to Atom package a hint: Don't tag releases yourself, or update the version manually, `apm publish` will take care of that for you, for more details on that see [here](https://atom.io/docs/latest/hacking-atom-package-word-count#publish-your-package)

dirk-thomas commented:
@steelbrain Thanks, the permissions look good now.

@Arcanemagus Thanks for the hint.
